presto-hive-metastore/src/main/java/com/facebook/presto/hive/metastore/thrift/DynamicHiveCluster.java

```diff
@@ -0,0 +1,93 @@
+import static java.util.Objects.requireNonNull;
+
+/**
+ * HiveCluster implementation to handle fetching of metastore URIs dynamically
+ */
+public class DynamicHiveCluster
+        implements HiveCluster
+{
+    private final HiveMetastoreClientFactory clientFactory;
+    private final String metastoreUsername;
+    private final HttpClient httpClient;
+    private final Request request;
+    private final MetastoreUriFetcher metastoreUriFetcher;
+
+    @Inject
+    public DynamicHiveCluster(DynamicMetastoreConfig config, HiveMetastoreClientFactory clientFactory, @ForDynamicHiveCluster HttpClient httpClient, MetastoreUriFetcher metastoreUriFetcher)
+    {
+        ThriftMetastoreHttpRequestDetails metaStoreDiscoveryUri = requireNonNull(config.getMetastoreDiscoveryUri(), "metaStoreDiscoveryUri object to capture http request details is null");
+        this.httpClient = requireNonNull(httpClient, "httpClient object is null");
+        this.clientFactory = requireNonNull(clientFactory, "clientFactory is null");
+        this.metastoreUriFetcher = requireNonNull(metastoreUriFetcher, "metastoreUriFetcher is null");
+        this.request = createMetastoreDiscoveryRequest(metaStoreDiscoveryUri);
+        this.metastoreUsername = config.getMetastoreUsername();
+    }
+
+    @Override
+    public HiveMetastoreClient createMetastoreClient()
+            throws TException
+    {
+        URI uri = metastoreUriFetcher.getMetastoreUri(httpClient, request);
+        checkMetastoreUri(uri);
+        HostAndPort address = HostAndPort.fromParts(uri.getHost(), uri.getPort());
+        try {
+            HiveMetastoreClient client = clientFactory.create(address);
+
+            if (!isNullOrEmpty(metastoreUsername)) {
+                client.setUGI(metastoreUsername);
+            }
+            return client;
+        }
+        catch (TException e) {
+            throw new TException(String.format("Failed connecting to Hive metastore %s on port %s", address.getHost(), address.getPort()), e);
+        }
+    }
+
+    private static Request createMetastoreDiscoveryRequest(ThriftMetastoreHttpRequestDetails metaStoreDiscoveryUri)
+    {
+        Request.Builder requestBuilder = prepareGet()
+                .setUri(uriBuilderFrom(URI.create(metaStoreDiscoveryUri.getUrl())).build());
+
+        metaStoreDiscoveryUri.getHeaders().forEach(requestBuilder::setHeader);
+        return requestBuilder.build();
+    }
+
+    private static URI checkMetastoreUri(URI uri)
+    {
+        requireNonNull(uri, "metastoreUri is null");
+        String scheme = uri.getScheme();
+        checkArgument(!isNullOrEmpty(scheme), "metastoreUri scheme is missing: %s", uri);
+        checkArgument(scheme.equals("thrift"), "metastoreUri scheme must be thrift: %s", uri);
+        checkArgument(uri.getHost() != null, "metastoreUri host is missing: %s", uri);
+        checkArgument(uri.getPort() != -1, "metastoreUri port is missing: %s", uri);
+        return uri;
+    }
```

Comment:
Let's have our first commit to do a refactoring by moving this helper to `ThriftMetastoreUtil` from `StaticHiveCluster`

Reply:
Should I create a separate PR for this to be committed first?